near-sdk/src/types/account_id.rs

```diff
@@ -0,0 +1,174 @@
+use borsh::{BorshDeserialize, BorshSchema, BorshSerialize};
+use serde::{de, Deserialize, Serialize};
+use std::convert::TryFrom;
+use std::fmt;
+
+use crate::env::is_valid_account_id;
+
+/// Account identifier. Provides access to user's state.
+#[derive(
+    Debug,
+    Clone,
+    PartialEq,
+    PartialOrd,
+    Ord,
+    Eq,
+    BorshDeserialize,
+    BorshSerialize,
+    Serialize,
+    Hash,
+    BorshSchema,
+)]
+pub struct AccountId(String);
+
+impl AccountId {
+    /// Returns reference to the account ID bytes.
+    pub fn as_bytes(&self) -> &[u8] {
+        self.0.as_bytes()
+    }
+    /// Returns reference to the account ID string.
+    pub fn as_str(&self) -> &str {
```

Comment:
I am debating if we should just `impl Deref<str> for AccountId`.

Reply:
I was thinking the same, but I wanted to explore without it first

Reply:
Just to update on thoughts, I think that adding the Deref to string opens up for a much more difficult API to maintain and really doesn't benefit in this case. I was originally considering `Deref<String>` because of how ubiquitously `String` is used for the account ID type, but I think it might be best to just move away from this pattern.

Also not sure if trying to move away from `String` inner representation altogether in favour of just `Vec<u8>` to minimize contract size in the future, and I'm just trying to be careful

Reply:
That's what I had thought, but apparently, it had caused bloat in the past, unless I misunderstood some reasoning.

Reply:
>but apparently, it had caused bloat in the past, unless I misunderstood some reasoning.

I think this is a miscommunication / misunderstanding (but maybe misunderstanding on my side :) ). I believe that neither `String` nor `str` causes bloat in comparison to `Vec<u8>` or `[u8]`. What causes bloat is string *formatting* (eg, unwrapping results) and accidentally using unicodo-aware APIs (#458).  Switching to `Vec<u8>` won't help with the first (`Vec<u8>` implements Debug). It might margianaly help with the second, but the cost would be that you'd have to re-implement ascii `lo_lowercase` on bytes directly yourself.